include/ews/ews.hpp

```diff
@@ -9429,238 +9429,1440 @@ namespace ews
+                return "HomePhone";
+            case phone_number_key::home_phone_2:
+                return "HomePhone2";
+            case phone_number_key::isdn:
+                return "Isdn";
+            case phone_number_key::mobile_phone:
+                return "MobilePhone";
+            case phone_number_key::other_fax:
+                return "OtherFax";
+            case phone_number_key::other_telephone:
+                return "OtherTelephone";
+            case phone_number_key::pager:
+                return "Pager";
+            case phone_number_key::primary_phone:
+                return "PrimaryPhone";
+            case phone_number_key::radio_phone:
+                return "RadioPhone";
+            case phone_number_key::telex:
+                return "Telex";
+            case phone_number_key::ttytdd_phone:
+                return "TtyTddPhone";
+            default:
+                throw exception("Bad enum value");
+            }
+        }
+    }
+
+    inline bool operator==(const im_address& lhs, const im_address& rhs)
+    {
+        return (lhs.key_ == rhs.key_) && (lhs.value_ == rhs.value_);
+    }
+
+    class phone_number final
+    {
+    public:
+        phone_number(phone_number_key k, std::string val)
+            : key_(std::move(k)), value_(std::move(val))
+        {
+        }
+
+        static phone_number
+        from_xml_element(const rapidxml::xml_node<char>& node)
+        {
+            using rapidxml::internal::compare;
+
+            // <t:PhoneNumbers>
+            //  <Entry Key="AssistantPhone">0123456789</Entry>
+            //  <Entry Key="BusinessFax">9876543210</Entry>
+            // </t:PhoneNumbers>
+
+            EWS_ASSERT(compare(node.local_name(), node.local_name_size(),
+                               "Entry", std::strlen("Entry")));
+            auto key = node.first_attribute("Key");
+            EWS_ASSERT(key && "Expected attribute Key");
+            return phone_number(
+                str_to_key(std::string(key->value(), key->value_size())),
+                std::string(node.value(), node.value_size()));
+        }
+
+        std::string to_xml() const
+        {
+            std::stringstream sstr;
+            sstr << " <t:"
+                 << "PhoneNumbers"
+                 << ">";
+            sstr << " <t:Entry Key=";
+            sstr << "\"" << internal::enum_to_str(key_);
+            sstr << "\">";
+            sstr << get_value();
+            sstr << "</t:Entry>";
+            sstr << " </t:"
+                 << "PhoneNumbers"
+                 << ">";
+            return sstr.str();
+        }
+
+        phone_number_key get_key() const { return key_; }
+        const std::string& get_value() const EWS_NOEXCEPT { return value_; }
+
+    private:
+        phone_number_key key_;
+        std::string value_;
+        friend bool operator==(const phone_number&, const phone_number&);
+
+        static phone_number_key str_to_key(const std::string& keystring)
```

Comment:
This function should be a free-function and member of `internal` namespace instead of a static member-function of a class.

Here and other locations.